File: python/shoe_sorting_demo/execution.py
from __future__ import annotations

from typing import Callable, List, Sequence, Tuple

import numpy as np

from bajie_sdk import (
    BajieRobot,
    CameraType,
    DetectObjectsRequest,
    MissionStatus,
    MovePlanStep,
    NamedBBox,
    ObjectDetection,
    ObjectPose3D,
    OvdEndpoint,
    Quatf,
    RGBDViewWithPose,
    Vec3f,
)
try:
    from .matching import custom_match_obj_views, _quat_wxyz_to_rotmat
except ImportError:
    from matching import custom_match_obj_views, _quat_wxyz_to_rotmat


def _normalize_or_raise(v: np.ndarray, *, err_msg: str) -> np.ndarray:
    n = float(np.linalg.norm(v))
    if n < 1e-9:
        raise ValueError(err_msg)
    return (v / n).astype(np.float64, copy=False)


def _rotmat_to_quatf(rot_mat: np.ndarray) -> Quatf:
    m = np.asarray(rot_mat, dtype=np.float64).reshape(3, 3)
    decision = np.empty(4, dtype=np.float64)
    decision[0] = m[0, 0]
    decision[1] = m[1, 1]
    decision[2] = m[2, 2]
    decision[3] = m[0, 0] + m[1, 1] + m[2, 2]
    choice = int(np.argmax(decision))
    quat = np.empty(4, dtype=np.float64)

    if choice != 3:
        i = choice
        j = (i + 1) % 3
        k = (j + 1) % 3
        quat[i] = 1.0 - decision[3] + 2.0 * m[i, i]
        quat[j] = m[j, i] + m[i, j]
        quat[k] = m[k, i] + m[i, k]
        quat[3] = m[k, j] - m[j, k]
    else:
        quat[0] = m[2, 1] - m[1, 2]
        quat[1] = m[0, 2] - m[2, 0]
        quat[2] = m[1, 0] - m[0, 1]
        quat[3] = 1.0 + decision[3]

    norm = float(np.sqrt(np.dot(quat, quat)))
    if not np.isfinite(norm) or norm < 1e-15:
        return Quatf(0.0, 0.0, 0.0, 1.0)
    quat /= norm
    return Quatf(float(quat[0]), float(quat[1]), float(quat[2]), float(quat[3]))


def _front_edge_horizontal_x_axis(
    head_image: RGBDViewWithPose,
    front_edge: Sequence[Sequence[float]],
) -> np.ndarray:
    if len(front_edge) < 2:
        raise ValueError("front_edge 至少需要两个点")
    p0 = np.asarray(front_edge[0], dtype=np.float64)
    p1 = np.asarray(front_edge[-1], dtype=np.float64)
    if p0.shape[0] < 2 or p1.shape[0] < 2:
        raise ValueError("front_edge 点至少需要两个坐标分量")

    edge = p1[:2] - p0[:2]
    if float(np.linalg.norm(edge)) < 1e-9:
        raise ValueError("front_edge 两端点过近，无法构造方向")

    z_axis = np.array([0.0, 0.0, 1.0], dtype=np.float64)
    edge_3d = np.array([edge[0], edge[1], 0.0], dtype=np.float64)
    x_axis = _normalize_or_raise(np.cross(z_axis, edge_3d), err_msg="front_edge 退化，无法构造水平法向")

    cam_x = _quat_wxyz_to_rotmat(head_image.tf_goal.orientation)[:, 2]
    cam_x = np.array([cam_x[0], cam_x[1], 0.0], dtype=np.float64)
    cam_norm = float(np.linalg.norm(cam_x))
    if cam_norm >= 1e-9 and float(np.dot(x_axis, cam_x / cam_norm)) < 0.0:
        x_axis = -x_axis
    return x_axis


def _place_quat_from_front_edge(
    head_image: RGBDViewWithPose,
    front_edge: Sequence[Sequence[float]],
) -> Quatf:
    x_axis = _front_edge_horizontal_x_axis(head_image, front_edge)
    z_axis = np.array([0.0, 0.0, 1.0], dtype=np.float64)
    y_axis = np.cross(z_axis, x_axis)
    y_axis /= float(np.linalg.norm(y_axis))
    x_axis = np.cross(y_axis, z_axis)
    x_axis /= float(np.linalg.norm(x_axis))
    return _rotmat_to_quatf(np.column_stack([x_axis, y_axis, z_axis]))

def bbox_center(bbox:tuple[int, int, int, int]) -> Tuple[float, float]:
    """目标框几何中心 ``(u, v)``（像素，浮点）。"""
    x1, y1, x2, y2 = bbox
    return ((x1 + x2) * 0.5, (y1 + y2) * 0.5)


def _say(robot: BajieRobot, text: str, *, no_voice: bool = False) -> None:
    print(f"[voice] {text}", flush=True)
    if no_voice:
        return
    try:
        robot.eco_robotSpeech(text)
    except Exception:
        pass


def _mission_succeeded(status: MissionStatus) -> bool:
    return int(status.error_info.code) == 0


def _place_position_pixel_ray_ground_z0(
    head_image: RGBDViewWithPose,
    uv: Tuple[float, float],
) -> Vec3f:
    """无有效深度时的回退：由像素与 ``tf_goal`` 反算放置点，相机射线与父系 **z=0** 平面求交，高度取 0。

    与 ``pixel_to_point_3d`` 共用内参语义：相机系 ``(x,y,z)`` 中 ``z`` 沿光轴向前；父系为
    ``tf_goal.header.frame_id``，``tf_goal.position`` 为相机原点在父系中的位置。
    """
    k = head_image.camera_info_k
    u, v = float(uv[0]), float(uv[1])
    fx, fy = float(k[0, 0]), float(k[1, 1])
    cx, cy = float(k[0, 2]), float(k[1, 2])
    if min(fx, fy) <= 0:
        raise ValueError("_place_position_pixel_ray_ground_z0: invalid intrinsics (fx/fy)")
    d_cam = np.array([(u - cx) / fx, (v - cy) / fy, 1.0], dtype=np.float64)
    R = _quat_wxyz_to_rotmat(head_image.tf_goal.orientation)
    t = np.array(
        [
            head_image.tf_goal.position.x,
            head_image.tf_goal.position.y,
            head_image.tf_goal.position.z,
        ],
        dtype=np.float64,
    )
    d_parent = R @ d_cam
    dz = float(d_parent[2])
    if abs(dz) < 1e-9:
        raise ValueError("_place_position_pixel_ray_ground_z0: ray parallel to z=0 (dz≈0)")
    s = -float(t[2]) / dz
    if s < 0:
        raise ValueError("_place_position_pixel_ray_ground_z0: intersection behind camera (s<0)")
    p = t + s * d_parent
    return Vec3f(x=float(p[0]), y=float(p[1]), z=0.0)


def mission_retry(*, attempts: int = 3, op_name: str = "mission",
                  robot: BajieRobot | None = None, no_voice: bool = False,
                  op_cn: str = "操作",
                  fail_do: Callable[[], None] | None = None,
                  ) -> Callable[[Callable[[], MissionStatus]], MissionStatus]:
    """对无参、返回 ``MissionStatus`` 的调用做重试（装饰器工厂用法）。

    用法：``mission_retry(attempts=3, op_name="pick")(lambda: robot.foo(...))``。
    单次失败：``error_info.code != 0`` 或调用抛 ``Exception``，均重试；用尽则 ``RuntimeError``。
    """

    def run(fn: Callable[[], MissionStatus]) -> MissionStatus:
        last_failure = ""
        for attempt in range(1, attempts + 1):
            try:
                status = fn()
            except Exception as exc:
                last_failure = "attempt %s/%s exception: %s" % (attempt, attempts, exc)
                print(
                    "[mission_retry] %s %s/%s failed (exception): %s"
                    % (op_name, attempt, attempts, exc),
                    flush=True,
                )
                if attempt < attempts and robot:
                    _say(robot, f"{op_cn}未成功，正在重试，别着急", no_voice=no_voice)
                if fail_do and attempt < attempts:
                    try: fail_do()
                    except Exception: pass
                continue
            if _mission_succeeded(status):
                print(
                    "[mission_retry] %s ok (attempt %s/%s)" % (op_name, attempt, attempts),
                    flush=True,
                )
                return status
            last_failure = "attempt %s/%s error_info=%s" % (attempt, attempts, status.error_info)
            print(
                "[mission_retry] %s %s/%s failed (mission status)" % (op_name, attempt, attempts),
                flush=True,
            )
            if attempt < attempts and robot:
                _say(robot, f"{op_cn}未成功，正在重试，别着急", no_voice=no_voice)
            if fail_do and attempt < attempts:
                try: fail_do()
                except Exception: pass
        print("[mission_retry] %s giving up after %s attempts" % (op_name, attempts), flush=True)
        raise RuntimeError("%s failed after %s attempts: %s" % (op_name, attempts, last_failure))

    return run


def eco_pick_with_arm_review_head_image_ref(
    robot: BajieRobot,
    view: RGBDViewWithPose,
    object_bbox: NamedBBox,
    head_image_ref_bbox: tuple[int, int, int, int],
    *,
    timeout_sec: float = 100.0,
) -> MissionStatus:
    """臂对准（``eco_lookto``）→ 臂图重检 → 跨视匹配 → ``eco_pick``。

    ``head_image_ref_bbox`` 为链首参考框；``object_bbox.bbox`` 为当步规划 ``from``。二者相等时为
    链上首步，对头图 bbox 做 ``lookto``；否则用链首框中心提 3D 再 ``lookto``。裁切匹配始终用
    ``head_image_ref_bbox``。
    """
    if head_image_ref_bbox == object_bbox.bbox:  # 首次移动
        _ = robot.eco_lookto(
            object_bbox.bbox,
            view=view,
            camera_type=CameraType.ARM,
            timeout_sec=120,
        )
    else:  # 链上后续步（from 可能为虚拟，用链首 3D 对准）
        look_pose = ObjectPose3D(
            frame_id=view.tf_goal.header.frame_id,
            position=view.pixel_to_point_3d(bbox_center(head_image_ref_bbox), True, 5),
            box_length=Vec3f(0.12, 0.1, 0.05),
        )
        _ = robot.eco_lookto(look_pose, camera_type=CameraType.ARM, timeout_sec=120)
    arm_image = robot.eco_captureImages(CameraType.ARM)
    arm_image.save_images_to_local("arm")
    arm_items: list[ObjectDetection] = list(
        robot.eco_detect_objects(
            DetectObjectsRequest(
                rgb_image=arm_image.rgb_image.img,
                labels=[object_bbox.name] if object_bbox.name else ["object"],
                entry=OvdEndpoint.SHOE,
            )
        ).items
    )
    matched_bbox = custom_match_obj_views(view, head_image_ref_bbox,
                                          arm_image, arm_items, robot)
    if matched_bbox is None:
        print(f"[WARN] 跨视图匹配失败，跳过鞋 {head_image_ref_bbox}", flush=True)
        for i, item in enumerate(arm_items):
            print(f"  arm[{i}]: bbox={getattr(item, 'bbox', 'N/A')} name={getattr(item, 'name', 'N/A')}", flush=True)
        raise RuntimeError("跨视图匹配失败")
    bbox = NamedBBox(id=0, name="shoe", bbox=matched_bbox)
    bbox.name += "鞋"  # 启动鞋子抓取模式
    status = robot.eco_pick(arm_image, bbox, timeout_sec=timeout_sec)
    return status


def execute_pick_and_place(
    robot: BajieRobot,
    label: str,
    head_image: RGBDViewWithPose,
    move: MovePlanStep,
    front_edge: Sequence[Sequence[float]],
    *,
    head_image_ref_bbox: tuple[int, int, int, int],
    no_voice: bool = False,
) -> None:
    """单步：臂图跨视抓（内含 ``lookto``）→ 按规划放。

    ``head_image_ref_bbox`` 是规划链在**首帧头图**上归约出的参考框，``move.from_bbox`` 是规划在
    当前步给出的位置（多步时可能被模拟改写）。

    * **首次移动**（``head_image_ref_bbox == move.from_bbox``）：同一条链在头图上的头一步，用头图
      bbox 给 ``eco_lookto``，与未拆链的 SDK 行为一致。
    * **链上后续步**（与 ``from`` 不相等，含「同物体第二次及以后」）：当步 ``from`` 与头图不对应，
      用 **链首** ``head_image_ref_bbox`` 的像素中心提 3D 点再 ``lookto``，避免用虚拟 from 对准。

    上述对准与抓取均在 :func:`eco_pick_with_arm_review_head_image_ref` 内完成；裁切始终用链首
    ``head_image_ref_bbox``。抓取、放置各自最多尝试 3 次；若 SDK 直接抛异常或返回非成功
    ``MissionStatus``，均计为一次失败并重试，用尽后 ``RuntimeError``。
    """
    print(
        "[execute_pick_and_place] begin label=%r first_step=%s to_center=%s"
        % (label, head_image_ref_bbox == move.from_bbox, move.to_center),
        flush=True,
    )

    _say(robot, "开始抓取", no_voice=no_voice)
    _ = mission_retry(attempts=3, op_name="pick",
                      robot=robot, no_voice=no_voice, op_cn="抓取",
                      fail_do=lambda: robot.eco_cancelAllMissions())(
        lambda: eco_pick_with_arm_review_head_image_ref(
            robot,
            head_image,
            object_bbox=NamedBBox(name=label, bbox=move.from_bbox),
            head_image_ref_bbox=head_image_ref_bbox,
            timeout_sec=120.0,
        )
    )

    _say(robot, "抓取成功，开始放置", no_voice=no_voice)

    # 规划 to 中心 + 前沿定水平朝向，6D 放置（优先深度反投影，失败则射线 ∩ z=0）
    try:
        place_position = head_image.pixel_to_point_3d(move.to_center, True, 5)
    except ValueError:
        place_position = _place_position_pixel_ray_ground_z0(head_image, move.to_center)
        print(place_position, flush=True)
    place_pose = ObjectPose3D(
        frame_id=head_image.tf_goal.header.frame_id,
        position=place_position,
        orientation=_place_quat_from_front_edge(head_image, front_edge),
        box_length=Vec3f(0.12, 0.1, 0.05),
    )
    _ = mission_retry(attempts=3, op_name="place_3D",
                      robot=robot, no_voice=no_voice, op_cn="放置")(
        lambda: robot.eco_place_3D(place_pose, timeout_sec=120.0)
    )

    _say(robot, "放置成功", no_voice=no_voice)
    print("[execute_pick_and_place] done label=%r" % (label,), flush=True)

File: cpp/shoe_sorting_demo/execution.cpp
#include "execution.h"

#include <cmath>
#include <iostream>
#include <stdexcept>

#include <robot_sdk/math_utils.h>

#include "matching.h"
#include "pairing.h"

namespace shoe_sorting {
namespace {

static void Say(const std::string& text, bool no_voice = false) {
    std::cout << "[voice] " << text << std::endl;
    if (no_voice) return;
    robot_sdk::eco_Speech(text);
}

/// 相机射线与父系 z=0 平面求交（匹配 Python _place_position_pixel_ray_ground_z0）。
static robot_sdk::Position PlacePositionPixelRayGroundZ0(
    const robot_sdk::PoseStamped& tf_goal,
    const std::vector<double>& camera_info_k,
    double u, double v) {
  double fx = camera_info_k[0], fy = camera_info_k[4];
  double cx = camera_info_k[2], cy = camera_info_k[5];

  double d_cam_x = (u - cx) / fx;
  double d_cam_y = (v - cy) / fy;
  double d_cam_z = 1.0;

  std::vector<double> R = robot_sdk::QuatToRotMat(tf_goal.pose.orientation);
  double tx = static_cast<double>(tf_goal.pose.position.x);
  double ty = static_cast<double>(tf_goal.pose.position.y);
  double tz = static_cast<double>(tf_goal.pose.position.z);

  double d_parent_x = R[0] * d_cam_x + R[3] * d_cam_y + R[6] * d_cam_z;
  double d_parent_y = R[1] * d_cam_x + R[4] * d_cam_y + R[7] * d_cam_z;
  double d_parent_z = R[2] * d_cam_x + R[5] * d_cam_y + R[8] * d_cam_z;

  double s = -tz / d_parent_z;
  robot_sdk::Position pos;
  pos.x = static_cast<float>(tx + s * d_parent_x);
  pos.y = static_cast<float>(ty + s * d_parent_y);
  pos.z = 0.0f;
  return pos;
}

// ======== PickOnce（对应 Python eco_pick_with_arm_review_head_image_ref） ========

static void PickOnce(const robot_sdk::ImageQueryNotifyData& head_image,
                     const std::vector<int>& from_bbox,
                     const std::string& object_name,
                     const std::vector<int>& head_image_ref_bbox) {
  // 1. eco_lookto（首次步 / 后续步，对应 Python 的 bbox / pose 重载）
  bool is_first =
      head_image_ref_bbox.size() >= 4 && from_bbox.size() >= 4 &&
      head_image_ref_bbox[0] == from_bbox[0] &&
      head_image_ref_bbox[1] == from_bbox[1] &&
      head_image_ref_bbox[2] == from_bbox[2] &&
      head_image_ref_bbox[3] == from_bbox[3];
  {
    if (is_first) {
      robot_sdk::BboxItem fb;
      fb.id = 0;
      fb.left_up_x = static_cast<float>(from_bbox[0]);
      fb.left_up_y = static_cast<float>(from_bbox[1]);
      fb.right_down_x = static_cast<float>(from_bbox[2]);
      fb.right_down_y = static_cast<float>(from_bbox[3]);

      robot_sdk::GetPoseData gp;
      gp.bbox = {fb};
      gp.rgb_image = head_image.rgb_image;
      gp.depth_image = head_image.depth_image;
      gp.tf_map = head_image.tf_goal;

      robot_sdk::GetPoseNotifyData result;
      robot_sdk::ErrorCode ec = robot_sdk::eco_GetPose(gp, result);
      if (ec.code != 0 || result.pose_results.empty()) {
        throw std::runtime_error("GetPose 无结果: code=" + std::to_string(ec.code));
      }

      const auto& p = result.pose_results[0];
      robot_sdk::BoxData box;
      box.position = p.position;
      box.orientation = p.orientation;
      box.box_length = p.box_length;
      box.frame_id = p.frame_id;

      ec = robot_sdk::eco_LookTo(robot_sdk::LookToTarget::HAND, box);
      if (ec.code != 0) {
        std::cout << "    臂对准非致命: code=" << ec.code << "（继续）" << std::endl;
      }
    } else {
      robot_sdk::BoxData box;
      double u = (head_image_ref_bbox[0] + head_image_ref_bbox[2]) * 0.5;
      double v = (head_image_ref_bbox[1] + head_image_ref_bbox[3]) * 0.5;
      try {
        box.position = robot_sdk::PixelToPoint3D(head_image, u, v);
        box.box_length = {0.12f, 0.10f, 0.05f};
        box.frame_id = head_image.tf_goal.header.frame_id;
        robot_sdk::ErrorCode ec2 = robot_sdk::eco_LookTo(robot_sdk::LookToTarget::HAND, box);
        if (ec2.code != 0) {
          throw std::runtime_error("臂对准失败: code=" + std::to_string(ec2.code) +
                                   ", msg=" + ec2.msg);
        }
      } catch (const std::exception& e) {
        std::cout << "    后续步对准失败: " << e.what() << std::endl;
      }
    }
  }

  // 2. eco_captureImages(CameraType.ARM)
  std::cout << "    臂图拍照..." << std::endl;
  robot_sdk::ImageQueryNotifyData arm_image;
  {
    robot_sdk::ImageQueryData q;
    q.camera_type = robot_sdk::CameraType::ARM;
    robot_sdk::ThrowOnError(robot_sdk::eco_ImageQuery(q, arm_image), "臂图拍照");
  }

  // 3. eco_detect_objects(ARM, OVD SHOE)
  std::cout << "    OVD 检测臂图..." << std::endl;
  auto arm_items = VLMDetect(arm_image.rgb_image.data,
                             {object_name});
  if (arm_items.empty()) {
    throw std::runtime_error("臂图未检测到目标");
  }

  // 4. 自定义跨视图匹配（对应 Python custom_match_obj_views）
  robot_sdk::RobotPos rp;
  if (robot_sdk::eco_RobotPosition(rp).code != 0) {
    throw std::runtime_error("获取机器人位置失败，无法进行跨视图匹配");
  }
  auto matched_bbox = shoe_sorting::CustomMatchObjViews(
      head_image, head_image_ref_bbox, arm_image, arm_items, rp);
  if (matched_bbox.size() < 4) {
    throw std::runtime_error("跨视图匹配失败");
  }

  // 5. eco_pick（bbox.name += "鞋" 触发鞋子抓取模式）
  std::string pick_name = object_name + "鞋";
  std::cout << "    eco_pick: name=" << pick_name << std::endl;

  robot_sdk::BoundingBox bbox;
  bbox.name = pick_name;
  bbox.left_up_x = static_cast<float>(matched_bbox[0]);
  bbox.left_up_y = static_cast<float>(matched_bbox[1]);
  bbox.right_down_x = static_cast<float>(matched_bbox[2]);
  bbox.right_down_y = static_cast<float>(matched_bbox[3]);

  robot_sdk::AccurateGrabData data;
  data.rgb_image = arm_image.rgb_image;
  data.depth_image = arm_image.depth_image;
  data.tf_goal = arm_image.tf_goal;
  data.bbox = bbox;

  robot_sdk::ThrowOnError(robot_sdk::eco_AccurateGrab(data), "抓取");
}

}  // namespace

// ======== ExecutePickAndPlace（对应 Python execute_pick_and_place） ========

void ExecutePickAndPlace(const robot_sdk::ImageQueryNotifyData& head_image,
                         const robot_sdk::ShoesIntentStep& step,
                         const std::vector<std::vector<double>>& front_edge,
                         const std::string& object_name,
                         const std::vector<int>& head_image_ref_bbox,
                         bool no_voice) {
  // ---- 1. 抓取（3 次重试，对应 Python mission_retry(attempts=3, op_name="pick")） ----
  Say("开始抓取", no_voice);                                     // 节点 E1
  std::cout << "    抓取: [" << step.from_bbox[0] << "," << step.from_bbox[1]
            << "," << step.from_bbox[2] << "," << step.from_bbox[3] << "]..."
            << std::endl;
  for (int attempt = 1;; ++attempt) {
    try {
      PickOnce(head_image, step.from_bbox, object_name, head_image_ref_bbox);
      break;
    } catch (const std::exception& e) {
      std::cout << "    抓取尝试 " << attempt << "/3 失败: " << e.what() << std::endl;
      if (attempt < 3) {
        robot_sdk::eco_StopAll(10*1000);
        Say("抓取未成功，正在重试，别着急", no_voice);            // 节点 E2
      }
      if (attempt >= 3) throw;
    }
  }
  Say("抓取成功，开始放置", no_voice);                           // 节点 E3
  std::cout << "    抓取成功" << std::endl;

  // ---- 2. 放置（匹配 Python 放置部分） ----
  double u = (step.to_bbox[0] + step.to_bbox[2]) * 0.5;
  double v = (step.to_bbox[1] + step.to_bbox[3]) * 0.5;

  robot_sdk::Position place_pos;
  try {
    place_pos = robot_sdk::PixelToPoint3D(head_image, u, v);
  } catch (const std::exception&) {
    place_pos = PlacePositionPixelRayGroundZ0(
        head_image.tf_goal, head_image.camera_info_k, u, v);
    std::cout << "    深度回退: 射线 ∩ z=0 → (" << place_pos.x << ", "
              << place_pos.y << ", " << place_pos.z << ")" << std::endl;
  }

  robot_sdk::Orientation orientation =
      robot_sdk::PlaceQuatFromFrontEdge(head_image.tf_goal, front_edge);
  robot_sdk::BoxLength box_length{0.12f, 0.10f, 0.05f};
  const std::string& frame_id = head_image.tf_goal.header.frame_id;

  for (int attempt = 1;; ++attempt) {
    std::cout << "    放置尝试 " << attempt << "/3: (" << place_pos.x << ", "
              << place_pos.y << ", " << place_pos.z << ")..." << std::endl;

    robot_sdk::AccuratePlaceData data;
    data.position = place_pos;
    data.orientation = orientation;
    data.box_length = box_length;
    data.frame_id = frame_id;

    robot_sdk::ErrorCode ec = robot_sdk::eco_AccuratePlace(data);
    if (ec.code == 0) break;

    std::string err = "放置失败: code=" + std::to_string(ec.code) + ", msg=" + ec.msg;
    std::cout << "    放置尝试 " << attempt << "/3 失败: " << err << std::endl;
    if (attempt < 3)
      Say("放置未成功，正在重试，请稍候", no_voice);              // 节点 E4
    if (attempt >= 3) throw std::runtime_error(err);
  }
  Say("放置成功", no_voice);                                     // 节点 E5
  std::cout << "    放置成功" << std::endl;
}

}  // namespace shoe_sorting
